brin: table AM API for block traversal

soumyadeep2007 opened the pull request:
This PR introduces a new table AM API for unifying code for block traversal used in BRIN indexes. To ensure that the changes don't break existing functionality, we had to bring back specific-range-summarization for heap tables. Also, we extended and improved the tests significantly. Please review commits individually.

This also fixes https://github.com/greenplum-db/gpdb/issues/14843.

Co-authored-by: Ashwin Agrawal <aashwin@vmware.com>

ashwinstar commented on src/backend/access/brin/brin.c:
```diff
@@ -520,151 +449,147 @@ bringetbitmap(IndexScanDesc scan, Node **bmNodeP)
-		if (RelationIsAppendOptimized(heapRel))
+		BlockNumber startblknum = sequences[i].startblknum;
+		BlockNumber endblknum = sequences[i].startblknum + sequences[i].nblocks;
+		for (heapBlk = startblknum; heapBlk < endblknum; heapBlk += opaque->bo_pagesPerRange)
```
One of the thought is to avoid intending the code and keep it upstream way (we have avoided in some places in GPDB to indent to avoid future merge conflicts, plus also makes it easier when comparing files against upstream to easily spot GPDB additions). I am not sure which way is better will leave it to you.

soumyadeep2007 replied:
Ah found an [example](https://github.com/greenplum-db/gpdb/commit/c1c97da95594db21dfd7956dd11729eb713d9114#diff-42bd5bf8f4f55d6314bf0238b70f037153dd4d2cf1f821028e36d79463b9bf35R266-R268).

Will do, it will definitely help I think! That's a great suggestion!

soumyadeep2007 replied:
Done!

ashwinstar commented on src/backend/access/brin/brin.c:
```diff
@@ -520,151 +449,147 @@ bringetbitmap(IndexScanDesc scan, Node **bmNodeP)
+				BlockNumber pageno;
+
+				for (pageno = heapBlk;
+					 pageno <= Min(endblknum, heapBlk + opaque->bo_pagesPerRange) - 1;
```
again know not ideal but if continue to use `nblocks` instead of `endblknum` feel can avoid change related to upstream here.

soumyadeep2007 replied:
Understand that, but it makes the code very hard to read and reason about, as nblocks only makes sense if we started from block 0, which we won't be most of the time in AO case.

ashwinstar commented on src/backend/access/brin/brin.c:
```diff
@@ -1484,107 +1409,123 @@ brinsummarize(Relation index, Relation heapRel, BlockNumber pageRange,
+
+		startBlk = (pageRange / pagesPerRange) * pagesPerRange;
+		endBlk = Min(sequences[0].nblocks, startBlk + pagesPerRange);
+		if (startBlk > Min(sequences[0].nblocks, startBlk + pagesPerRange))
```
seems simpler to use `startBlk > endBlk` here

soumyadeep2007 replied:
Done

ashwinstar commented on src/backend/access/brin/brin.c:
```diff
@@ -1484,107 +1409,123 @@ brinsummarize(Relation index, Relation heapRel, BlockNumber pageRange,
 	{
-		BrinTuple  *tup;
-		OffsetNumber off;
+		if (!specificRange)
```
feel might be better to pull `if (pageRange == BRIN_ALL_BLOCKRANGES)` if..else block here inside loop to gain clarity as startBlk initialization for both if and else case will happen here, That way can avoid the variable as well `specificRange`

soumyadeep2007 replied:
Hmm could do that, but wanted to keep the initialization code outside the loop as consistent with upstream as possible. See the if-else above outside the loop.

I could remove the extra variable and replace `if (!specificRange)` with `if (pageRange == BRIN_ALL_BLOCKRANGES)`.

Let me know your preference.

ashwinstar replied:
It will still be outside `for (; startBlk < endBlk; startBlk += pagesPerRange)` upstream loop, just would be inside GPDB specific loop where it seems right place to initialize the start and end range for inner loop. Initialization of startBlk and endBlk much further out for specificRange case gets little confusing to read and follow along.

soumyadeep2007 replied:
Sounds good. Will do. Yes that will make it much clearer. Its all a very careful balance between readability and upstream alignment! Thanks so much for all the advice!

soumyadeep2007 replied:
Done

soumyadeep2007 commented on src/test/regress/expected/brin_aocs_optimizer.out:
```diff
@@ -463,20 +469,70 @@ UPDATE brintest_aocs SET textcol = '' WHERE textcol IS NOT NULL;
-          0 |     1 |      102 |     1
-          0 |     2 |      102 |     1
-(6 rows)
+-- We should have two segment files.
```
Yeah don't need the column number. Anyway, have gotten rid of this in the 3rd commit, which updates the tests.

soumyadeep2007 commented on src/include/access/appendonlytid.h:
```diff
@@ -56,9 +56,24 @@ typedef struct AOTupleId
+ *
+ * append-optimized table logically has 128 segment files. The highest 7 bits
+ * of the logical Tid represent the segment file number. So, segment file number
+ * with zero after is the start block number in a segment file.
```
Will do, that was brought over from the function: segnoGetCurrentAosegStart

soumyadeep2007 replied:
Done

ashwinstar commented on src/backend/access/brin/brin.c:
```diff
@@ -1484,107 +1409,123 @@ brinsummarize(Relation index, Relation heapRel, BlockNumber pageRange,
+
+		Assert(RelationIsHeap(heapRel));
+
+		numSequences = 1; /* set up to loop only once below */
```
`table_relation_get_block_sequences` should have done this for us anyways. Maybe wish to assert on it?

soumyadeep2007 replied:
Done

haolinw commented:
`table_relation_get_block_sequences` is missed in GetTableAmRoutine().

soumyadeep2007 commented on src/backend/access/aocs/aocsam_handler.c:
```diff
@@ -1924,6 +1924,61 @@ aoco_relation_size(Relation rel, ForkNumber forkNumber)
+		pfree(seginfos);
+	}
+
+	*numSequences = nsegs;
```
Done!

soumyadeep2007 commented on src/include/access/tableam.h:
```diff
@@ -610,6 +638,11 @@ typedef struct TableAmRoutine
+	 * See table_relation_get_block_sequences()
+	 */
+	BlockSequence *(*relation_get_block_sequences) (Relation rel,
+													int *numSequences);
```
Fixed

soumyadeep2007 commented on src/include/access/tableam.h:
```diff
@@ -1739,6 +1772,17 @@ table_relation_size(Relation rel, ForkNumber forkNumber)
+ */
+static inline BlockSequence *
+table_relation_get_block_sequences(Relation rel,
+								   int *numSequences)
```
Fixed

soumyadeep2007 commented on src/backend/access/heap/heapam_handler.c:
```diff
@@ -2094,6 +2094,27 @@ heapam_relation_size(Relation rel, ForkNumber forkNumber)
+	blockSequence->startblknum = 0;
+	blockSequence->nblocks = RelationGetNumberOfBlocks(rel);
+
+	*numRanges = 1;
```
Done

soumyadeep2007 commented:
I have addressed most of the feedback, some more to go. I have found some more issues, fixing those. Marking the PR as draft for now. Will reopen once work is finalized.

soumyadeep2007 commented on src/backend/access/brin/brin.c:
```diff
@@ -458,8 +390,14 @@ bringetbitmap(IndexScanDesc scan, Node **bmNodeP)
+	int         	numSequences;
+	BlockSequence 	*sequences;
+	MemoryContext	oldcontext;
+	MemoryContext 	rangesContext = AllocSetContextCreate(CurrentMemoryContext,
```
Right, got rid of it.

soumyadeep2007 commented on src/test/regress/expected/brin_ao.out:
```diff
@@ -445,34 +446,87 @@ INSERT INTO brintest_ao SELECT
+BEGIN
+  LOOP
+    INSERT INTO brin_ao_summarize VALUES (1) RETURNING ctid INTO curtid;
+    EXIT WHEN curtid > tid '(33554434, 0)';
```
Done as a high-level comment in summarize section.

soumyadeep2007 commented:
> table_relation_get_block_sequences is missed in GetTableAmRoutine().
Fixed.

soumyadeep2007 commented:
Ready for review again. All tests pass, assert fixed, more test cases added. Also ran src/test/modules/brin, which passes with the exception of `01_workitems.pl` (which needs user-table autovac to be enabled).

ashwinstar commented on src/backend/access/appendonly/appendonlyam_handler.c:
```diff
@@ -1952,6 +1952,31 @@ appendonly_relation_get_block_sequences(Relation rel,
+	Oid 			segrelid;
+
+	segno = AOSegmentGet_segno(blkNum);
+	GetAppendOnlyEntryAuxOids(rel, &segrelid, NULL, NULL, NULL, NULL);
```
we should be able to use the relcache now to get segrelid for all of these calls instead of fetching them from catalog.

soumyadeep2007 replied:
GetAppendOnlyEntryAuxOids is now a simple getter function after the relcache changes. We are fetching the results from relcache indirectly, as now we just peer into `rel->rd_appendonly`, which was populated as part of relation_open().

ashwinstar replied:
Interesting okay.  Do we really need a function call for the same, directly can reference or maybe a macro is fine. Though not for this PR.

I get worried about these kind of references from relcache
```
    aoForm = rel->rd_appendonly;  <----- current code seems safe though in future some code gets added in-between here and if recache invalidation happens in-between then there is scope for dereferencing invalid/stale pointer. So, generally better to reference the relcache variables directly without storing the pointers for them

    if (segrelid != NULL)                                                                                                                   
        *segrelid = aoForm->segrelid;   
```
Refer this bug for cdbpolicy which used to cause crashes https://github.com/greenplum-db/gpdb/pull/5248. I haven't checked if rd_appendonly gets rebuild during invalidation or not.

soumyadeep2007 replied:
Hmm yes it does get rebuilt during invalidation:

`RelationCacheInvalidate() -> RelationClearRelation() -> RelationBuildDesc() -> RelationInitAppendOnlyInfo()`

soumyadeep2007 replied:
>     aoForm = rel->rd_appendonly;  <----- current code seems safe though in future some code gets added in-between here and if recache invalidation happens in-between then there is scope for dereferencing invalid/stale pointer. So, generally better to reference the relcache variables directly without storing the pointers for them

cc @huansong.

It might be best to replace GetAppendOnlyEntryAuxOids function with a macro bearing the same name and signature.

huansong replied:
> It might be best to replace GetAppendOnlyEntryAuxOids function with a macro bearing the same name and signature.

Makes sense. Probably also `GetAppendOnlyEntry`. Will do

ashwinstar commented on src/backend/access/brin/brin.c:
```diff
@@ -1295,10 +1301,18 @@ summarize_range(IndexInfo *indexInfo, BrinBuildState *state, Relation heapRel,
+	 * tables, heapBlk need not be a multiple of bs_pagesPerRange.
 	 */
-	Assert(heapBlk % state->bs_pagesPerRange == 0);
+	AssertImply(RelationIsHeap(heapRel), heapBlk % state->bs_pagesPerRange == 0);
```
actually, only for starting blocknumber of the range for AO this is not valid assert but still valid for all other block numbers. Not sure if its easy to make that distinction here. Possibly via `AOSegmentGet_startHeapBlock()`

soumyadeep2007 replied:
Minimal repro of failure with attached patch.

```sql
postgres=# create table foo(i int) using ao_row;         
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'i' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
CREATE TABLE
postgres=# create index on foo using brin(i) with (pages_per_range=3);
CREATE INDEX
postgres=# DO $$                                         
DECLARE curtid tid;
BEGIN
  LOOP
    INSERT INTO foo VALUES (1) RETURNING ctid INTO curtid;
    EXIT WHEN curtid > tid '(33554435, 0)';
  END LOOP;
END;
$$;
DO
postgres=# SELECT brin_summarize_new_values('foo_i_idx');
ERROR:  Unexpected internal error (assert.c:44)  (seg1 slice1 192.168.0.148:7003 pid=1993827) (assert.c:44)
DETAIL:  AssertImply failed("!(!(((heapRel)->rd_rel->relam == 2) || heapBlk != ((((((heapBlk)) / (1 << 25))) << 25))) || (heapBlk % state->bs_pagesPerRange == 0))", File: "brin.c", Line: 1310)
[relax.patch.txt](https://github.com/greenplum-db/gpdb/files/10844270/relax.patch.txt)

CONTEXT:  SQL function "brin_summarize_new_values" statement 1
```

`summarize_range` is called with `heapBlk = 33554432` for the first range and then `heapBlk = 33554435` for the second range. `33554435` is not a multiple of 3.

Thus, this assert does not apply.

ashwinstar replied:
Ahh... my bad, yes if starting range is not aligned then further blocks won't be either...

ashwinstar commented on src/backend/access/appendonly/appendonlyam_handler.c:
```diff
@@ -1952,6 +1952,31 @@ appendonly_relation_get_block_sequences(Relation rel,
+	segno = AOSegmentGet_segno(blkNum);
+	GetAppendOnlyEntryAuxOids(rel, &segrelid, NULL, NULL, NULL, NULL);
+
+	sequence->startblknum = AOSegmentGet_startHeapBlock(segno);
```
not sure how much worth it is but this part of code is common with `appendonly_relation_get_block_sequences()`, maybe can add inline common function to have it in one place.

soumyadeep2007 replied:
Done! Added macro

ashwinstar commented on src/backend/access/table/tableamapi.c:
```diff
@@ -93,6 +93,7 @@ GetTableAmRoutine(Oid amhandler)
 	Assert(routine->index_validate_scan != NULL);
 
 	Assert(routine->relation_size != NULL);
+	Assert(routine->relation_get_block_sequences != NULL);
```
need `relation_get_block_sequence` as well here

soumyadeep2007 replied:
Done!

ashwinstar commented on src/backend/access/brin/brin.c:
```diff
@@ -520,12 +443,25 @@ bringetbitmap(IndexScanDesc scan, Node **bmNodeP)
-	for (heapBlk = 0; heapBlk < nblocks; heapBlk += opaque->bo_pagesPerRange)
+	BlockNumber startblknum = sequences[i].startblknum;
+	BlockNumber endblknum = sequences[i].startblknum + sequences[i].nblocks;
+	for (heapBlk = startblknum; heapBlk < endblknum; heapBlk += opaque->bo_pagesPerRange)
```
Just for reference, not the problem of this PR but in one of BRIN related discussions upstream, this loop can be infinite due to overflow. Chances of that happening for HEAP table is super low, though for AO it does exist specially for 128th segfile and very high pages_per_range. Refer https://www.postgresql.org/message-id/b8a4e04c-c091-056c-a379-11d35c7b2d8d%40enterprisedb.com. So, would be good to keep eye and once that aspect lands in PostgreSQL at least bring it to GPDB to be safe.

soumyadeep2007 replied:
Oh dear. Yes we need to fix this. I like Tomas' approach - it's dead simple.

I will add this in a dedicated PR.

There are other issues too that upstream hasn't yet addressed.
Excerpt from one of the patches:

```
+	 * XXX This needs rethinking to make it work with large indexes with more
+	 * ranges than we can fit into memory (work_mem/maintenance_work_mem).
+	 */
+	for (heapBlk = 0; heapBlk < nblocks; heapBlk += pagesPerRange)
```

ashwinstar commented on src/backend/access/appendonly/appendonlyam_handler.c:
```diff
@@ -1721,9 +1721,7 @@ appendonly_relation_get_block_sequences(Relation rel,
-		sequences[i].nblocks = lastSequence / AO_MAX_TUPLES_PER_HEAP_BLOCK;
-		if (lastSequence % AO_MAX_TUPLES_PER_HEAP_BLOCK > 0)
-			sequences[i].nblocks += 1;
+		FastSequenceGetNumHeapBlocks(lastSequence, &sequences[i].nblocks);
```
I was thinking `ReadLastSequence()` and `AOSegmentGet_startHeapBlock()` can be delegated to the common macro/function as that's all duplicate code. So, essentially it can populate both nblocks and startblknum in same common place for a sequence.

soumyadeep2007 replied:
Done!

ashwinstar commented on src/test/isolation2/input/uao/brin.source:
```diff
@@ -0,0 +1,47 @@
+END; /* in func */
+$$; /* in func */
+
+SELECT gp_wait_until_triggered_fault('summarize_last_partial_range', 1, dbid)
```
we wish to wait before performing insert to make sure the code path is covered, else can be flaky behavior depending on the timing of overlap of summarize and insert.

soumyadeep2007 replied:
Done, great catch!
